Mathlib/GroupTheory/SpecificGroups/Alternating/Centralizer.lean

```diff
@@ -0,0 +1,301 @@
+
+* Finally, `Equiv.Perm.OnCycleFactors.kerφ_le_alternating_iff`
+  establishes on which iff-condition the centralizer of an even permutation
+  is contained in `alternatingGroup α`.
+
+TODO :
+Deduce the formula for the cardinality of the centralizers
+and conjugacy classes in `alternatingGroup α`.
+-/
+
+open Equiv Finset Function MulAction
+
+variable {α : Type*} [Fintype α] [DecidableEq α] (g : Perm α)
+
+namespace Equiv.Perm.OnCycleFactors
+
+variable (k : Perm (fixedPoints g))
+    (v : (c : g.cycleFactorsFinset) → Subgroup.zpowers (c : Perm α))
+
+theorem sign_kerParam_apply_apply :
+    sign (kerParam g ⟨k, v⟩) = sign k * ∏ c, sign (v c : Perm α) := by
+  rw [kerParam, MonoidHom.noncommCoprod_apply, ← Prod.fst_mul_snd ⟨k, v⟩, Prod.mk_mul_mk, mul_one,
+    one_mul, map_mul, sign_ofSubtype, Finset.univ_eq_attach, mul_right_inj, ← MonoidHom.comp_apply,
+    Subgroup.noncommPiCoprod, MonoidHom.comp_noncommPiCoprod _, MonoidHom.noncommPiCoprod_apply,
+    univ_eq_attach, noncommProd_eq_prod]
+  simp
+
+theorem cycleType_kerParam_apply_apply :
+    cycleType (kerParam g ⟨k, v⟩) = cycleType k + ∑ c, cycleType (v c : Perm α) := by
+  let U := (Finset.univ : Finset { x // x ∈ g.cycleFactorsFinset }).toSet
+  have hU : U.Pairwise fun i j ↦ (v i : Perm α).Disjoint (v j : Perm α) := fun c _ d _ h ↦ by
+    obtain ⟨m, hm⟩ := (v c).prop
+    obtain ⟨n, hn⟩ := (v d).prop
+    simp only [← hm, ← hn]
+    apply Disjoint.zpow_disjoint_zpow
+    apply cycleFactorsFinset_pairwise_disjoint g c.prop d.prop
+    exact Subtype.coe_ne_coe.mpr h
+  rw [kerParam, MonoidHom.noncommCoprod_apply, ← Prod.fst_mul_snd ⟨k, v⟩, Prod.mk_mul_mk, mul_one,
+    one_mul, univ_eq_attach, Disjoint.cycleType (disjoint_ofSubtype_noncommPiCoprod g k v),
+    Subgroup.noncommPiCoprod_apply, Disjoint.cycleType_noncommProd hU, univ_eq_attach]
+  exact congr_arg₂ _ cycleType_ofSubtype rfl
+
+variable {g} in
+theorem odd_of_centralizer_le_alternatingGroup (h : Subgroup.centralizer {g} ≤ alternatingGroup α) :
+    ∀ i ∈ g.cycleType, Odd i := by
+  intro i hi
+  rw [cycleType_def g, Multiset.mem_map] at hi
+  obtain ⟨c, hc, rfl⟩ := hi
+  rw [← Finset.mem_def] at hc
+  suffices sign c = 1 by
+    rw [IsCycle.sign _, neg_eq_iff_eq_neg, ← Int.units_ne_iff_eq_neg] at this
+    · rw [← Nat.not_even_iff_odd, comp_apply]
+      exact fun h ↦ this h.neg_one_pow
+    · rw [mem_cycleFactorsFinset_iff] at hc
+      exact hc.left
+  apply h
+  rw [Subgroup.mem_centralizer_singleton_iff]
+  exact Equiv.Perm.self_mem_cycle_factors_commute hc
+
+end Equiv.Perm.OnCycleFactors
+
+namespace AlternatingGroup
+
+open Nat Equiv.Perm.OnCycleFactors Equiv.Perm
+
+private theorem of_cycleType_aux (m : Multiset ℕ) :
+    map (Embedding.subtype fun x ↦ x ∈ alternatingGroup α)
+      (univ.filter fun g : alternatingGroup α => (g : Equiv.Perm α).cycleType = m) =
+      if Even (m.sum + Multiset.card m)
+        then Finset.univ.filter fun g : Equiv.Perm α => g.cycleType = m
+        else ∅ := by
```

Comment:
I think this is a bit easier on the eye if we use our `Finset` builder notation and lean a bit more heavily the elaborator:
```suggestion
    ({g | (g : Perm α).cycleType = m} : Finset (alternatingGroup α)).map (Embedding.subtype _) =
      if Even (m.sum + m.card) then ({g | g.cycleType = m} : Finset (Perm α)) else ∅ := by
```